src/Compilers/VisualBasic/Portable/Language/LanguageFeature.vb

```diff
@@ -0,0 +1,291 @@
+        ''' Feature maybe enabled explicitly with a Feature Flag or enabled implicitly via a Language Version.
+        ''' </remarks>
+        ''' <param name="feature">Language feature to check is available.</param>
+        ''' <param name="options">The parse options being used.</param>
+        <Extension>
+        Friend Function IsAvailable(feature As Feature, options As VisualBasicParseOptions) As Boolean
+            Return feature.IsInLanguageVersion(options) OrElse feature.CheckFeatures(options)
+        End Function
+
+        ''' <summary>
+        ''' Check to see if a language <paramref name="feature"/> is available with this <paramref name="compilation"/>.
+        ''' </summary>
+        ''' <param name="feature">Language feature to check is available.</param>
+        ''' <param name="compilation">The <see cref="VisualBasicCompilation"/> being used.</param>
+        <Extension>
+        Friend Function IsAvailable(feature As Feature, compilation As VisualBasicCompilation) As Boolean
+            Return feature.IsInLanguageVersion(compilation)
+        End Function
+
+        ''' <summary>
+        ''' Check to see if a language <paramref name="feature"/> is available with the <paramref name="options"/> being used.
+        ''' If unavailable the function return the node with an unavailable diagnostic attached to <paramref name="node"/>.
+        ''' </summary>
+        ''' <returns>If <see cref="feature"/> is not available, returns node with unavailable diagnostic attached to it.</returns>
+        ''' <param name="node">The node to attach the potential diagnostic (Feature Unavailable).</param>
+        ''' <param name="feature">Language feature to check is available.</param>
+        ''' <param name="options">The parse options being used.</param>
+        <Extension>
+        Friend Function CheckFeatureAvailability(Of TNode As Syntax.InternalSyntax.VisualBasicSyntaxNode)(node As TNode, feature As Feature, options As VisualBasicParseOptions) As TNode
+            Return If(feature.IsAvailable(options), node, node.ReportFeatureUnavailable(feature, options))
+        End Function
+
+        ''' <summary>
+        ''' Is a language <paramref name="feature"/> unavailable to use with these <paramref name="options"/>.
+        ''' </summary>
+        ''' <returns>
+        ''' Return a feature unavailable <see cref="Diagnostic"/> at <paramref name="location"/>, otherwise nothing.
+        ''' </returns>
+        ''' <param name="feature">Language feature to check is available.</param>
+        ''' <param name="options">The parse options being used.</param>
+        ''' <param name="location">The location to report the diagnostic.</param>
+        <Extension>
+        Friend Function CheckFeatureAvailable(feature As Feature, options As VisualBasicParseOptions, location As Location) As Diagnostic
+            Return If(feature.IsAvailable(options), Nothing, ReportFeatureUnavailable(feature, options, location))
+        End Function
```

Comment:
Please rename to `CheckFeatureAvailability` for consistency with the method above (and the current `Parser` method), and include a `DiagnosticBag` parameter: `Function CheckFeatureAvailability(feature As Feature, options As VisualBasicParserOptions, location As Location, diagnostics As DiagnosticBag) As Boolean`

Reply:
Renamed but I'm leaving signature as shown. so code isn't side effecting. Adding to `DiagnosticBag` is left for caller.

Reply:
It is reasonable for this method to add to a `DiagnosticBag`. Callers are using this method specifically for that purpose. Please restore the `DiagnosticBag` parameter.

Reply:
@cston The updated version of `CheckFeatureAvailability` is now small enough to be eligible for in-lining by the JIT.